codex/rng.nim

```diff
@@ -32,10 +32,8 @@ const randMax = 18_446_744_073_709_551_615'u64
 proc rand*(rng: Rng, max: Natural): int =
   if max == 0: return 0
 
-  var x: uint64
   while true:
-    var rng = rng[]
-    hmacDrbgGenerate(rng, addr x, csize_t(sizeof(x)))
+    let x = rng[].generate(uint64)
```

Comment:
Just wanting to understand this better. Looking at `func generate` in [bearssl/rand](https://github.com/status-im/nim-bearssl/blob/master/bearssl/rand.nim), the first parameter is `var HmacDrbgContext` for all impls.

So when caller passes `rng[]` it's effectively like passing a `var`? It's probably an aspect of the calling conventions I didn't fully appreciate.

The reason I ask is that `type Rng* = ref HmacDrbgContext` so I was thinking of `rng[]` as deref'ing from `ref HmacDrbgContext` to just `HmacDrbgContext`.

Reply:
`var HmacDrbgContext` = a HmacDrbgContext you can modify
`ref HmacDrbgContext` = a managed pointer to a HmacDrbgContext you can modify

When you use `[]`, you just dereference that pointer to a value you can modify
_However_, if you assign it to a variable it will be copied to that variable (no longer a "shared reference")

`var` parameters have a different semantic, a `var` parameter is actually a non-nil `ref` under the hood